ga4gh/datamodel/sequenceAnnotations.py

```diff
@@ -0,0 +1,432 @@
+        :return: dictionary representing a feature object,
+            or None if no match is found.
+        """
+        sql = ("SELECT * FROM FEATURE WHERE id = ?")
+        query = self._dbconn.execute(sql, (featureId,))
+        ret = query.fetchone()
+        if ret is None:
+            return None
+        return sqliteBackend.sqliteRow2Dict(ret)
+
+
+# Problem: FeatureIds from GFF3 often contain colons,
+# and the CompoundId mechanism reserves colons for its special separator
+# character. Thus the following pair of functions that mangles and
+# de-mangles featureId's to allow them to work in compoundIds.
+def _decolonize(featureId):
+    """
+    Replaces colons in input with CompoundId-safe character
+    :param featureId: string potentially with colons in it
+    :return: same string with colons replaced with safe character
+    """
+    return featureId.replace(':', ';')
+
+
+def _recolonize(mangled_featureId):
+    """
+    Replaces CompoundId-safe character in input with colons
+    :param featureId: string potentially with replacement character
+    :return: same string with colons in place.
+    """
+    return mangled_featureId.replace(';', ':')
+
+
+class AbstractFeatureSet(datamodel.DatamodelObject):
+    """
+    A set of sequence features annotations
+    """
+    compoundIdClass = datamodel.FeatureSetCompoundId
+
+    def __init__(
+            self, parentContainer, localId, referenceSetId=None):
+        super(AbstractFeatureSet, self).__init__(parentContainer, localId)
+        self._referenceSetId = referenceSetId
+        self._name = localId
+        self._sourceUri = ""
+        self._attributes = protocol.Attributes()
+
+    def toProtocolElement(self):
+        """
+        Returns the representation of this FeatureSet as the corresponding
+        ProtocolElement.
+        """
+        gaFeatureSet = protocol.FeatureSet()
+        gaFeatureSet.id = self.getId()
+        gaFeatureSet.datasetId = self.getParentContainer().getId()
+        gaFeatureSet.referenceSetId = self._referenceSetId
+        gaFeatureSet.name = self._name
+        gaFeatureSet.sourceUri = self._sourceUri
+        gaFeatureSet.attributes = self._attributes
+        return gaFeatureSet
+
+    def getCompoundIdForFeatureId(self, featureId):
+        """
+        Returns server-style compound ID for a GFF3 style feature ID.
+
+        :param featureId: string feature ID as reported in GFF3
+        :return: string representing ID for the specified GA4GH protocol
+            Feature object in this FeatureSet. If None is passed in,
+            an empty string is passed out, as required by the protocol.
+        """
+        if featureId is None:
+            # This can happen if no feature ID is present in the database,
+            # as can be the case with (absent) parent IDs.
+            return ""
+        compoundId = datamodel.FeatureCompoundId(
+            self.getCompoundId(), _decolonize(featureId))
```

Comment:
Although in general I believe encoding identifiers this way is an anti-pattern instead of adding and removing separators you might consider base64 encoding/decoding again. Although it could lead to extremely lengthy identifiers you will have good guarantees your characters are safe.